File: game-of-life/kotlin-arrow-game-of-life/src/test/kotlin/GamePlayTest.kt
import GamePlay.calculateNextGeneration
import org.hamcrest.CoreMatchers.equalTo
import org.junit.Assert.assertThat
import org.hamcrest.core.Is.`is`
import org.junit.Test

class GamePlayTest {

    @Test
    fun `a single cell matrix should die`() {
        val nextGenerationMatrix = calculateNextGeneration(listOf(
                listOf(Alive)
        )).unsafeRunSync()

        assertThat(nextGenerationMatrix, `is`(equalTo(listOf(listOf(Dead)) as Matrix)))
    }

    @Test
    fun `a 3x3 matrix with a single alive cell in the middle`() {
        val nextGenerationMatrix = calculateNextGeneration(listOf(
                listOf(Dead, Dead, Dead),
                listOf(Dead, Alive, Dead),
                listOf(Dead, Dead, Dead)
        )).unsafeRunSync()

        assertThat(nextGenerationMatrix, `is`(equalTo(listOf(
                listOf(Dead, Dead, Dead),
                listOf(Dead, Dead, Dead),
                listOf(Dead, Dead, Dead)
        ) as Matrix)))
    }

    @Test
    fun `a 3x3 matrix where cells has 2 neighbours and it is alive or have 3 neighbours and they are dead`() {
        val nextGenerationMatrix = calculateNextGeneration(listOf(
                listOf(Alive, Alive, Dead),
                listOf(Dead, Alive, Dead),
                listOf(Dead, Dead, Dead)
        )).unsafeRunSync()

        assertThat(nextGenerationMatrix, `is`(equalTo(listOf(
                listOf(Alive, Alive, Dead),
                listOf(Alive, Alive, Dead),
                listOf(Dead, Dead, Dead)
        ) as Matrix)))
    }

    @Test
    fun `a 4x4 matrix where cells has 2 neighbours and it is alive or have 3 neighbours and they are dead`() {
        val nextGenerationMatrix = calculateNextGeneration(listOf(
                listOf(Alive, Alive, Dead, Alive),
                listOf(Dead, Alive, Alive, Dead),
                listOf(Dead, Dead, Dead, Alive),
                listOf(Dead, Dead, Alive, Alive)
        )).unsafeRunSync()

        assertThat(nextGenerationMatrix, `is`(equalTo(listOf(
                listOf(Alive, Alive, Dead, Dead),
                listOf(Alive, Alive, Dead, Alive),
                listOf(Dead, Alive, Dead, Alive),
                listOf(Dead, Dead, Alive, Alive)
        ) as Matrix)))
    }

}

File: game-of-life/typescript-fpts-game-of-life/__test__/GamePlay.test.ts
import {calculateNextGeneration} from "../src/GamePlay";
import {alive, dead} from "../src/Data";

describe('GamePlay', () => {
    it('a single cell matrix should die', async () => {
        const matrix = await calculateNextGeneration([[alive()]])();

        expect(matrix).toEqual([[dead()]])
    });

    it('a 3x3 matrix with a single alive cell in the middle', async () => {
        const matrix = await calculateNextGeneration([
            [dead(), dead(), dead()],
            [dead(), alive(), dead()],
            [dead(), dead(), dead()]
        ])();

        expect(matrix).toEqual([
            [dead(), dead(), dead()],
            [dead(), dead(), dead()],
            [dead(), dead(), dead()]
        ])
    });

    it('a 3x3 matrix where cells has 2 neighbours and it is alive or have 3 neighbours and they are dead', async () => {
        const matrix = await calculateNextGeneration([
            [alive(), alive(), dead()],
            [dead(), alive(), dead()],
            [dead(), dead(), dead()]
        ])();

        expect(matrix).toEqual([
            [alive(), alive(), dead()],
            [alive(), alive(), dead()],
            [dead(), dead(), dead()]
        ])
    });

    it('a 4x4 matrix where cells has 2 neighbours and it is alive or have 3 neighbours and they are dead', async () => {
        const matrix = await calculateNextGeneration([
            [alive(), alive(), dead(), alive()],
            [dead(), alive(), alive(), dead()],
            [dead(), dead(), dead(), alive()],
            [dead(), dead(), alive(), alive()]
        ])();

        expect(matrix).toEqual([
            [alive(), alive(), dead(), dead()],
            [alive(), alive(), dead(), alive()],
            [dead(), alive(), dead(), alive()],
            [dead(), dead(), alive(), alive()]
        ])
    });
});
